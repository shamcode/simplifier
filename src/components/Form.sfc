<template>
    {% import Input from './Input.sfc' %}

    <form
        :onsubmit={{this$._onSubmit}}
        :ref={{$.formNode}}
    >
        <Input
            id="input-text"
            placeholder="(a && b) || (!a && !b)"
            name="expression"
        />
        <button type="submit" tabindex="2">Simplify!</button>
    </form>
</template>

<script>

    function Form( options ) {
        const onSubmit = $();

        const state = options( {
            [ onSubmit ]() {}
        } );

        this$._onSubmit = ( evt ) => {
            const data = new FormData( this$.formNode );
            state[ onSubmit ]( evt, {
                expression: data.get( 'expression' )
            } );
        };
    }

    export default Component( Template, Form );
</script>
