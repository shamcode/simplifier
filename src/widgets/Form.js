import { options } from 'sham-ui';
import FormTemplate from './Form.sht';

export default class Form extends FormTemplate {
    @options onSubmit() {}

    _onSubmit( evt ) {
        const data = new FormData( this.formNode );
        this.options.onSubmit( evt, {
            expression: data.get( 'expression' )
        } );
    }
}
