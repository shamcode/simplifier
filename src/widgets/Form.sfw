<template>
    {% import Input from './Input.sfw' %}

    <form
        :onsubmit={{this._onSubmit.bind( this )}}
        :ref="formNode"
    >
        <Input
            id="input-text"
            tabindex="1"
            placeholder="(a && b) || (!a && !b)"
            name="expression"
        />
        <button type="submit" tabindex="2">Simplify!</button>
    </form>
</template>

<script>
    import { options } from 'sham-ui';

    /**
     * @property {Element} formNode
     */
    export default class Form extends Template {
        @options onSubmit() {}

        _onSubmit( evt ) {
            const data = new FormData( this.formNode );
            this.options.onSubmit( evt, {
                expression: data.get( 'expression' )
            } );
        }
    }
</script>