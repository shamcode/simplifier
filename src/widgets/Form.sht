{% import Input from './Input' %}

<form>
    <Input
            id="input-text"
            tabindex="1"
            placeholder="(a && b) || (!a && !b)"
            name="expresion"
    />
    <button type="submit" tabindex="2">Simplify!</button>
</form>
