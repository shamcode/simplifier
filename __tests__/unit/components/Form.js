import Form from '../../../src/components/Form.sfc';
import { ref, onsubmit } from 'sham-ui-directives';
import renderer from 'sham-ui-test-helpers';

it( 'renders correctly', () => {
    const meta = renderer( Form, {}, {
        directives: {
            ref,
            onsubmit
        }
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'onSubmit options', () => {
    expect.assertions( 3 );
    const fn = jest.fn();

    const meta = renderer( Form, {
        onSubmit: fn
    }, {
        directives: {
            ref,
            onsubmit
        }
    } );

    meta.component.update();
    meta.ctx.container.querySelector( 'input' ).value = 'a && b';
    meta.ctx.container.querySelector( 'form' ).dispatchEvent( new Event( 'submit' ) );

    const json = meta.toJSON();
    delete json.Options.onSubmit;
    expect( json ).toMatchSnapshot();
    expect( fn.mock.calls ).toHaveLength( 1 );
    expect( fn.mock.calls[ 0 ][ 1 ] ).toEqual( { expression: 'a && b' } );
} );


it( 'onSubmit props', () => {
    expect.assertions( 3 );
    const fn = jest.fn();

    const meta = renderer( Form, {}, {
        directives: {
            ref,
            onsubmit
        }
    } );

    meta.component.update( {
        onSubmit: fn
    } );
    meta.ctx.container.querySelector( 'input' ).value = 'a && b';
    meta.ctx.container.querySelector( 'form' ).dispatchEvent( new Event( 'submit' ) );

    const json = meta.toJSON();
    delete json.Options.onSubmit;
    expect( json ).toMatchSnapshot();
    expect( fn.mock.calls ).toHaveLength( 1 );
    expect( fn.mock.calls[ 0 ][ 1 ] ).toEqual( { expression: 'a && b' } );
} );
