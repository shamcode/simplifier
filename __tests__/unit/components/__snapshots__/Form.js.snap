// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onSubmit options 1`] = `
Object {
  "Constructor": "Form",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "directives": Object {
      "onsubmit": [Function],
      "ref": [Function],
    },
    "types": Array [],
  },
  "html": "
<form>
  <input id=\\"input-text\\" placeholder=\\"(a &amp;&amp; b) || (!a &amp;&amp; !b)\\" tabindex=\\"1\\" name=\\"expression\\">
  <!--Input-->
  <button type=\\"submit\\" tabindex=\\"2\\">Simplify!</button>
</form>
",
}
`;

exports[`onSubmit props 1`] = `
Object {
  "Constructor": "Form",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "directives": Object {
      "onsubmit": [Function],
      "ref": [Function],
    },
    "types": Array [],
  },
  "html": "
<form>
  <input id=\\"input-text\\" placeholder=\\"(a &amp;&amp; b) || (!a &amp;&amp; !b)\\" tabindex=\\"1\\" name=\\"expression\\">
  <!--Input-->
  <button type=\\"submit\\" tabindex=\\"2\\">Simplify!</button>
</form>
",
}
`;

exports[`renders correctly 1`] = `
Object {
  "Constructor": "Form",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "directives": Object {
      "onsubmit": [Function],
      "ref": [Function],
    },
    "onSubmit": [Function],
    "types": Array [],
  },
  "html": "
<form>
  <input id=\\"input-text\\" placeholder=\\"(a &amp;&amp; b) || (!a &amp;&amp; !b)\\" tabindex=\\"1\\" name=\\"expression\\">
  <!--Input-->
  <button type=\\"submit\\" tabindex=\\"2\\">Simplify!</button>
</form>
",
}
`;
